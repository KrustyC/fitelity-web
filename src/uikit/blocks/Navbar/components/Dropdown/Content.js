import React from 'react';
import styled, { css, keyframes } from 'styled-components';

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }
`;

const ContentWrapper = styled.div`
  min-width: 160px;
  z-index: 1000;

  &:before {
    content: '';
    position: absolute;
    background: #fff;
    width: 20px;
    height: 20px;
    top: -3px;
    z-index: -1;
    left: 50%;
    margin-left: -10px;
    transform: rotate(45deg);
  }

  ${({ theme }) => css`
    @media (min-width: ${theme.sizes.md}) {
      position: absolute;
      animation: ${fadeIn} 0.4s ease-in-out;
      margin-top: 5px;
      box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
    }
  `}
`;

const Content = props => <ContentWrapper {...props} />;
Content.displayName = 'NavbarDropdownContent';

export default Content;
