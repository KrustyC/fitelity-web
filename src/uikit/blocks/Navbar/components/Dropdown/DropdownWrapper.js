import styled, { css } from 'styled-components';

const DropdownWrapper = styled.div`
  display: flex;
  justify-content: center;

  ${({ theme }) => css`
    @media (min-width: ${theme.sizes.md}) {
      max-width: 200px;
    }

    @media (max-width: ${theme.sizes.md}) {
      width: 100%;
      flex-direction: column;
    }
  `}
`;

export default DropdownWrapper;
