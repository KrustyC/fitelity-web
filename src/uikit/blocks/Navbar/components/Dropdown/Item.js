import React from 'react';
import styled, { css } from 'styled-components';
import { Link } from 'react-router-dom';

const ItemWrapper = styled(Link)`
  float: none;
  padding: 12px 16px;
  text-decoration: none;
  display: block;
  text-align: left;
  background-color: white;
  width: 100%;

  ${({ theme }) => css`
    &:hover {
      background-color: ${theme.colors.primary};
      text-decoration: none;
    }
  `}
`;

const Item = props => <ItemWrapper {...props} />;
Item.displayName = 'NavbarDropdownItem';

export default Item;
