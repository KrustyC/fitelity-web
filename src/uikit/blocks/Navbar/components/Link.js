import React from 'react';
import styled, { css } from 'styled-components';
import BaseLink from 'uikit/elements/Link';

const StyledLink = styled(BaseLink)`
  display: flex;
  text-decoration: none;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  z-index: 100;
  height: 100%;
  min-width: 60px;

  &:hover,
  &:focus {
    text-decoration: none;
  }

  ${({ theme }) => css`
    padding-left: 20px;
    padding-right: 20px;

    &.active {
      background: ${theme.colors.primary};
    }

    &:hover,
    &:focus {
      background: ${theme.colors.primary};
    }

    @media (min-width: ${theme.sizes.md}) {
      &::after {
        content: '';
        position: absolute;
        width: 100%;
        transform: scaleX(0);
        height: 2px;
        bottom: 0;
        left: 0;
        background: ${theme.colors.primary};
        transform-origin: bottom right;
        transition: transform 0.25s ease-out;
      }

      &:hover::after {
        transform: scaleX(1);
        transform-origin: bottom left;
      }

      &.active::after {
        transform: scaleX(1);
        transform-origin: bottom left;
      }
    }

    @media (max-width: ${theme.sizes.md}) {
      flex: 1;
      justify-content: flex-start;
      min-height: 50px;
    }
  `}

  span {
    display: none;
    ${({ theme }) => css`
      @media (max-width: ${theme.sizes.md}) {
        display: block;
        text-transform: capitalize;
        margin-left: 10px;
      }
    `}
  }
`;

const Link = props => <StyledLink type="nav" {...props} />;
Link.displayName = 'NavbarLink';

export default Link;
