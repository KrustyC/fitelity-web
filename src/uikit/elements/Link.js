import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import {
  Link as ReactRouterLink,
  NavLink as ReactRouterNavLink,
} from 'react-router-dom';

import Button from 'uikit/blocks/Button';

const Link = styled(ReactRouterLink)`
  color: ${({ theme }) => theme.colors.accent};
  text-decoration: none;
  cursor: pointer;

  ${({ font }) =>
    font === 'default' &&
    css`
      font-family: ${({ theme }) => theme.fonts.default};
    `} ${({ font }) =>
    font === 'serif' &&
    css`
      font-family: ${({ theme }) => theme.fonts.serif};
    `}

  &:hover, &:focus {
    color: ${({ theme }) => theme.util.darkenOnHover(theme.colors.accent)};
    text-decoration: underline;
  }

  &:hover,
  &:focus {
    color: ${({ theme }) => theme.util.darkenOnActive(theme.colors.accent)};
    text-decoration: underline;
  }
`;

const NavLink = Link.withComponent(ReactRouterNavLink);
const ALink = Link.withComponent('a');

const LinkElement = props => {
  let { component } = props;
  const { children, to, ...rest } = props;

  if (String(to).indexOf('http') > -1) {
    component = 'a';
  }

  switch (component) {
    case 'button':
      return (
        <Link to={to} as={Button} {...rest}>
          {children}
        </Link>
      );
    case 'nav':
      return (
        <NavLink to={to} {...rest}>
          {children}
        </NavLink>
      );
    case 'a':
      return (
        <ALink {...rest} href={to} target="_blank">
          {children}
        </ALink>
      );
    default:
      return (
        <Link to={to} {...rest}>
          {children}
        </Link>
      );
  }
};

LinkElement.propTypes = {
  children: PropTypes.any.isRequired,
  component: PropTypes.oneOf(['button', 'index']),
  font: PropTypes.oneOf(['default', 'serif']),
};

LinkElement.defaultProps = {
  component: null,
};

export default LinkElement;
