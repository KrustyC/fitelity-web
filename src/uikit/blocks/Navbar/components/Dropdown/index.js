import React, { Component } from 'react';
import PropTypes from 'prop-types';

import DropdownWrapper from './DropdownWrapper';
import Content from './Content';
import Item from './Item';

class NavbarDropdown extends Component {
  static propTypes = {
    children: PropTypes.any.isRequired,
  };

  constructor(props) {
    super(props);

    this.state = {
      expanded: false,
    };

    this.dropdownRef = React.createRef();
  }

  componentWillMount() {
    document.addEventListener('mousedown', this.handleClick, false);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClick, false);
  }

  handleClick = e => {
    e.preventDefault();
    if (!this.dropdownRef.current.contains(e.target)) {
      // user is clicking outside
      this.setState({ expanded: false });
    }
  };

  toggleExpand = e => {
    e.preventDefault();
    const { expanded } = this.state;
    this.setState({ expanded: !expanded });
  };

  render() {
    const { children } = this.props;
    const { expanded } = this.state;

    const [button, ...items] = children;
    return (
      <DropdownWrapper innerRef={this.dropdownRef}>
        {React.cloneElement(button, {
          onClick: this.toggleExpand,
        })}
        {expanded && <Content>{items}</Content>}
      </DropdownWrapper>
    );
  }
}

NavbarDropdown.Item = Item;
NavbarDropdown.displayName = 'NavbarDropdown';

export default NavbarDropdown;
