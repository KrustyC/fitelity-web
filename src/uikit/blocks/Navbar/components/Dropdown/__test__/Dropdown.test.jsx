import React from 'react'

import mountWithContexts from 'App/helpers/mountWithContexts'
import renderWithContexts from 'App/helpers/renderWithContexts'
import Dropdown from '../'
import Link from '../../Link'

describe('Navbar Dropdown', () => {
  it('toggle the dropdown list when clicking on its first link', () => {
    const wrapper = mountWithContexts(
      <Dropdown>
        <Link to="/uikit/dropdown">Dropdown</Link>
        <Dropdown.Item to="/uikit/dropdown/1">Item One</Dropdown.Item>
        <Dropdown.Item to="/uikit/dropdown/2">Item Two</Dropdown.Item>
        <Dropdown.Item to="/uikit/dropdown/3">Item Three</Dropdown.Item>
      </Dropdown>
    )

    const dropdown = wrapper.childAt(0)
    const button = dropdown.find(Link)

    expect(wrapper.find(Dropdown.Item)).toHaveLength(0)

    button.simulate('click')
    expect(wrapper.find(Dropdown.Item)).toHaveLength(3)

    button.simulate('click')
    expect(wrapper.find(Dropdown.Item)).toHaveLength(0)
  })

  it('renders correctly', () => {
    const tree = renderWithContexts(
      <Dropdown>
        <Link to="/uikit/dropdown">Dropdown</Link>
        <Dropdown.Item to="/uikit/dropdown/1">Item One</Dropdown.Item>
        <Dropdown.Item to="/uikit/dropdown/2">Item Two</Dropdown.Item>
        <Dropdown.Item to="/uikit/dropdown/3">Item Three</Dropdown.Item>
      </Dropdown>
    )
    expect(tree).toMatchSnapshot()
  })
})
