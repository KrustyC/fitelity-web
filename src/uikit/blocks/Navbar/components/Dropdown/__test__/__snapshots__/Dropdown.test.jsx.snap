// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Dropdown renders correctly 1`] = `
<div
  className="sc-bdVaJa kZriCf"
>
  <a
    aria-current={false}
    className="sc-htoDjs jQZqDc sc-bZQynM iKYpnw"
    href="/uikit/dropdown"
    onClick={[Function]}
    style={undefined}
    type="nav"
  >
    Dropdown
  </a>
</div>
`;
